dove/src/index/mod.rs

```diff
@@ -265,16 +256,129 @@ impl Index {
         self.deps_roots
             .iter()
             .map(PathBuf::from)
-            .filter(|path| {
-                let com = path
-                    .components()
-                    .filter_map(|el| el.as_os_str().to_str())
-                    .collect::<Vec<&str>>();
-                com.iter()
-                    .find_position(|el| el == &&".external")
-                    .and_then(|(pos, _)| com.get(pos + 1))
-                    .map_or(false, |el| el.starts_with("git_"))
-            })
+            .filter(|path| is_external_git(path))
             .for_each(remove);
     }
+
+    /// difference in indexes
+    pub fn diff<'a>(&'a self, index: &'a Index) -> Vec<Diff<'a>> {
+        let cur_paths = &self.deps_roots;
+        let oth_paths = &index.deps_roots;
+
+        let mut diff: Vec<Diff> = cur_paths
+            .iter()
+            .map(|path| {
+                if oth_paths.contains(path) {
+                    Diff::Unchanged(path)
+                } else {
+                    Diff::Deleted(path)
+                }
+            })
+            .collect();
+
+        diff.extend(oth_paths.iter().filter_map(|path| {
+            if cur_paths.contains(path) {
+                None
+            } else {
+                Some(Diff::Added(path))
+            }
+        }));
+
+        diff
+    }
+}
+
+/// difference in indexes
+#[derive(Debug, Eq, PartialEq)]
+pub enum Diff<'a> {
+    /// Added the path to the index
+    Added(&'a str),
+    /// The path was removed from the index
+    Deleted(&'a str),
+    /// No change
+    Unchanged(&'a str),
+}
+
+impl<'a> Diff<'a> {
+    /// Check if the path to delete is marked
+    pub fn is_deleted(&self) -> bool {
+        matches!(self, Diff::Deleted(_))
+    }
+
+    /// convert to path
+    pub fn to_pathbuf(&self) -> PathBuf {
+        let path = match self {
+            Diff::Deleted(path) => path,
+            Diff::Added(path) => path,
+            Diff::Unchanged(path) => path,
+        };
+        PathBuf::from(path)
+    }
+}
+
+/// Is the directory of the downloaded repository
+fn is_external_git(path: &Path) -> bool {
+    let com = path
```

Comment:
What is `com`? Please use full names instead of abbreviations. It also seems that collecting to vector is unnecessary here.